src/Features/LanguageServer/Protocol/Handler/RequestExecutionQueue.QueueItem.cs

```diff
@@ -85,40 +122,118 @@ internal partial class RequestExecutionQueue
+                    if (cancellationToken.IsCancellationRequested)
+                    {
+                        _logger.TraceInformation($"{MethodName} - Canceled while in queue");
+                        this.Metrics.RecordCancellation();
+
+                        _completionSource.SetCanceled();
+                        return;
+                    }
+
+                    // If we weren't able to get a corresponding context for this request (for example, we
+                    // couldn't map a doc request to a particular Document, or we couldn't find an appropriate
+                    // Workspace for a global operation), then just immediately complete the request with a
+                    // 'null' response.  Note: the lsp spec was checked to ensure that 'null' is valid for all
+                    // the requests this could happen for.  However, this assumption may not hold in the future.
+                    // If that turns out to be the case, we could defer to the individual handler to decide
+                    // what to do.
+                    if (context == null)
+                    {
+                        _logger.TraceWarning($"Could not get request context for {MethodName}");
+                        this.Metrics.RecordFailure();
+
+                        _completionSource.SetResult(default);
+                        return;
+                    }
+
+                    var result = await _handler.HandleRequestAsync(_request, context.Value, cancellationToken).ConfigureAwait(false);
+
                     this.Metrics.RecordSuccess();
+
+                    _completionSource.SetResult(result);
                 }
-                catch (OperationCanceledException)
+                catch (OperationCanceledException ex)
                 {
+                    // Record logs + metrics on cancellation.
                     _logger.TraceInformation($"{MethodName} - Canceled");
                     this.Metrics.RecordCancellation();
-                    throw;
+
+                    // Only cancel the task completion source so the caller (client) can react.
+                    // We don't need cancellation exceptions bubbling up to the request queue.
+                    _completionSource.TrySetCanceled(ex.CancellationToken);
                 }
                 catch (Exception ex)
                 {
+                    // Record logs and metrics on the exception.
                     _logger.TraceException(ex);
                     this.Metrics.RecordFailure();
-                    throw;
+
+                    // Pass the exception to the task completion source, so the caller (client) can react
+                    _completionSource.SetException(ex);
+
+                    // Also allow the exception to flow back to the request queue to handle as appropriate
+                    throw new InvalidOperationException($"Error handling '{MethodName}' request: {ex.Message}", ex);
```

Comment:
this is def a bit... interesting.  not saying it's wrong.  but i always find ti a bit suspect . it would be good to have a deeper comment on why this is needed.  for example, since we did `_completionSource.SetException(ex);` this somewhat implies that someone is not observing the TCS itself (but still is observing the direction exception thrown).  

I wouldn't mind an actual case listed where that happens.  note: i get this is a move, so we should keep even if you can't give a case.

Reply:
This part is fairly important.

The `_completionSource` represents the task completion as the LSP client sees it, so we of course send the exception result back to the client.

However, that `completionSource` doesn't represent the result of the call inside the queue processing.  By throwig the exception here as well, the queue processing in `ProcessQueueAsync` also sees the exception and decides (depending on if the request was mutating or not) how to handle the exception (in the case of mutating, it shuts down the server, for non-mutating it just reports a non-fatal).